# Application_Voyage - Kayak

*jedha Nov. 2023*  
Mélanie Beraud _ melanie.beraud@gmail.com

### API or website to scape:
GPS coordinates: https://nominatim.org/  
Weather: https://openweathermap.org/appid   
Hotels: booking.com

## Imports
### Import Librairies

In [122]:
# import
## import for API use
import requests
import pandas as pd

# import for graphs
import matplotlib.pyplot as plt
##import seaborn as sns

#import plotly.io as pio
import plotly.express as px
#import plotly.offline as pyo
#from plotly import graph_objects as go
#from plotly.subplots import make_subplots
#pio.renderers.default = "iframe_connected" # permet d'avoir de l'interactivité
#pyo.init_notebook_mode(True)

#https://plotly.com/python/renderers/
#These renderers write figures out as standalone HTML files and then display iframe elements that reference these HTML files.
#With the iframe renderers, the figure data are stored in the individual HTML files rather than in the notebook itself, resulting in a smaller notebook size.
#renderer="notebook_connected": This saves a few megabytes in notebook size, but an Internet connection is required in order to display figures that are rendered this way.




print('Libraires imported')

Libraires imported


https://www.youtube.com/watch?v=2SRRyKTl2IY

### Import fichiers et données importantes

In [4]:
top_35 = ["Mont Saint Michel", "St Malo", "Bayeux", "Le Havre", "Rouen", 
          "Paris", "Amiens", "Lille", "Strasbourg", "Chateau du Haut Koenigsbourg",
           "Colmar", "Eguisheim", "Besancon", "Dijon", "Annecy", 
           "Grenoble", "Lyon", "Gorges du Verdon", "Bormes les Mimosas", "Cassis", 
           "Marseille", "Aix en Provence", "Avignon", "Uzes", "Nimes", 
           "Aigues Mortes", "Saintes Maries de la mer", "Collioure", "Carcassonne", "Ariege", 
           "Toulouse", "Montauban", "Biarritz", "Bayonne", "La Rochelle"]

print('List \'top_35\' imported')

List 'top_35' imported


## GPS coordinates

https://nominatim.openstreetmap.org/

    /search - search OSM objects by name or type
    /reverse - search OSM object by their location
    /lookup - look up address details for OSM objects by their ID
    /status - query the status of the server
    /deletable - list objects that have been deleted in OSM but are held back in Nominatim in case the deletion was accidental
    /polygons - list of broken polygons detected by Nominatim
    /details - show internal details for an object (for debugging only)

In [17]:
# set city parameters
payload = {'city':'Paris', 'format':'json'}
#payload='Paris' #idem

# go fetch !
reponse = requests.get('https://nominatim.openstreetmap.org/search', params=payload)

print('ReponseStatus:\n', reponse) #200
print('Retour en texte:\n', reponse.text) # no gps coordinates
#print('Retour cookies:\n', reponse.cookies)
#print('Retour headers:\n', reponse.headers)

# write json
data_r = reponse.json()['adresstype'=='city']

# recover interesting data from json
coord = {'ville':data_r['name'], 'place_id': data_r['place_id'], 
         'lon':data_r['lon'], 
         'lat':data_r['lat'], 
         'lon/lat': data_r['lon'] + '/' + data_r['lat'], 
         'display_name':data_r['display_name']}
type(coord) #dict

ReponseStatus:
 <Response [200]>
Retour en texte:
 [{"place_id":83293737,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright","osm_type":"relation","osm_id":71525,"lat":"48.8534951","lon":"2.3483915","class":"boundary","type":"administrative","place_rank":12,"importance":0.8317101715588673,"addresstype":"city","name":"Paris","display_name":"Paris, Île-de-France, France métropolitaine, France","boundingbox":["48.8155755","48.9021560","2.2241220","2.4697602"]},{"place_id":83293355,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright","osm_type":"relation","osm_id":7444,"lat":"48.8588897","lon":"2.3200410217200766","class":"boundary","type":"administrative","place_rank":15,"importance":0.8317101715588673,"addresstype":"suburb","name":"Paris","display_name":"Paris, Île-de-France, France métropolitaine, France","boundingbox":["48.8155755","48.9021560","2.2241220","2.4697602"]}]


dict

{'ville': 'Paris',
 'place_id': 83293737,
 'lon': '2.3483915',
 'lat': '48.8534951',
 'lon/lat': '2.3483915/48.8534951',
 'display_name': 'Paris, Île-de-France, France métropolitaine, France'}

Example reponse.text :
[{"place_id":83293737,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright","osm_type":"relation","osm_id":71525,"lat":"48.8534951","lon":"2.3483915","class":"boundary","type":"administrative","place_rank":12,"importance":0.8317101715588673,"addresstype":"city","name":"Paris","display_name":"Paris, Île-de-France, France métropolitaine, France","boundingbox":["48.8155755","48.9021560","2.2241220","2.4697602"]},{"place_id":83293355,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright","osm_type":"relation","osm_id":7444,"lat":"48.8588897","lon":"2.3200410217200766","class":"boundary","type":"administrative","place_rank":15,"importance":0.8317101715588673,"addresstype":"suburb","name":"Paris","display_name":"Paris, Île-de-France, France métropolitaine, France","boundingbox":["48.8155755","48.9021560","2.2241220","2.4697602"]}]
Example data_r:
{'place_id': 83293737,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright',
 'osm_type': 'relation',
 'osm_id': 71525,
 'lat': '48.8534951',
 'lon': '2.3483915',
 'class': 'boundary',
 'type': 'administrative',
 'place_rank': 12,
 'importance': 0.8317101715588673,
 'addresstype': 'city',
 'name': 'Paris',
 'display_name': 'Paris, Île-de-France, France métropolitaine, France',
 'boundingbox': ['48.8155755', '48.9021560', '2.2241220', '2.4697602']}

 Example coord:
 {'ville': 'Paris',
 'place_id': 83293737,
 'lat': '48.8534951',
 'lon': '2.3483915',
 'display_name': 'Paris, Île-de-France, France métropolitaine, France'}

In [28]:
#USELESS TRANSFO FOR NOW
''' 
coord['lon/lat'] = coord['lon'] + '/' + coord['lat']
coord

df_coord = pd.DataFrame(coord, index = [1])
print(type(df_coord))
df_coord
'''

<class 'pandas.core.frame.DataFrame'>


ville  place_id        lon         lat               lon/lat  \
1  Paris  83293737  2.3483915  48.8534951  2.3483915/48.8534951   

                                        display_name  
1  Paris, Île-de-France, France métropolitaine, F...

In [ ]:
''' DO NOT RUN - MAPS tests using go

fig = px.scatter_geo(df_coord['lon/lat'],
                     name="Paris"
                     #, locations = df_coord['ville'],
                     #size=1#_ size of markers, "pop" is one of the columns of gapminder
                     )
plotly.graph_objects.Scattergeo
fig.show() # affiche le graph, mais pas la lat et la long


coord = {'lon': [2.3483915],'lat': [48.8534951]} #,'lon/lat': '2.3483915, 48.8534951'} #, 'city':'Paris','':
#coord = {'cluster': [2.3483915, 48.8534951]} doesn't work

fig = go.Scattermapbox(coord) 
type(fig) # en fait, c'est une espèce de dico que l'on créé... Pas du tout un graphique en lui même. Il faut ensuite l'appeler avec px ou autre
# car c'est un objet appelable dans une fonction px. Donc on doit ensuite le montrer soit avec la classe go.Figure (appeler en instance), soit avec px


#fig = Figure([data, layout, frames, skip_invalid])
fig = go.Figure(                                                            #si je le mets là, je n'ai pas à le mettre avec px ensuite...
    data = go.Scattermapbox(coord,                                           #data = go.Scattermapbox(coord['lat/lon'] #ne fonctionne pas avec lat/lon
    mode = 'markers'#, 
                                                                        #marker = go.scatter.Marker(color = 'city') # separation par couleurs
    ),
    layout = go.Layout(
        title = go.layout.Title(text = "MappingCities", x = 0.5), # centre
        #xaxis = go.layout.XAxis(title = 'Sepal width', range = [0, 5]),
        #yaxis = go.layout.YAxis(title = 'Sepal length', range = [0, 10])
    )
)
#fig
px.scatter_geo().show(renderer="notebook_connected")
fig.show(renderer="notebook_connected") # j'ai un titre et rien en dessous need the rendered of doesn't work. eful here because already in the import zone.
#px.scatter_geo(fig).show()  # j'ai une map du monde, mais elle est vide
#px.scatter(fig).show() # j'ai un scatter plot avec des valeurs type data, layout, frames (valeur du goScattermapbox)


DOESN'T WORK
coord = {'lon': [2.3483915],'lat': [48.8534951]}#,'city':['Paris']}

fig = go.Figure(                                                            #si je le mets là, je n'ai pas à le mettre avec px ensuite...
    data = go.Scattermapbox(coord,                                           #data = go.Scattermapbox(coord['lat/lon'] #ne fonctionne pas avec lat/lon
    mode = 'markers'#, 
                                                                        #marker = go.scatter.Marker(color = 'city') # separation par couleurs
    ),
    layout = go.Layout(
        title = go.layout.Title(text = "MappingCities", x = 0.5), # centre
        #xaxis = go.layout.XAxis(title = 'Sepal width', range = [0, 5]),
        #yaxis = go.layout.YAxis(title = 'Sepal length', range = [0, 10])
    )
)

fig3 = px.scatter_mapbox(data_frame = fig,
                         #text = 'city', already on the map
                         #title = 'Cities',
                         mapbox_style = 'carto-positron')   #,size=, mapbox_style='open-street-map' = very precise
fig3.show(renderer="notebook_connected") # nothing, not even a map

'''

In [182]:
coord = {'lon': 2.3483915,'lat': 48.8534951,'city':'Paris'}
df_coord = pd.DataFrame(coord, index=[1])
display(df_coord)

fig3 = px.scatter_mapbox(data_frame = df_coord,
                         lat = 'lat',
                         lon = 'lon',
                         #text = 'city', already on the map
                         mapbox_style='carto-positron')   #,size=, mapbox_style='open-street-map' = very precise
fig3.show(renderer="notebook_connected") # nothing, not even a map

lon        lat   city
1  2.348391  48.853495  Paris